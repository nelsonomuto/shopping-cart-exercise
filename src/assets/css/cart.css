.cart-item {
  display: flex;
  flex-flow: row wrap;
}
.cart-item-title {
  flex: 1 0 100%;
}
.cart-item-price {
  display: flex;
}
.cart-item-controls {
  display: flex;
}
.cart-item-update-qty-button {
  display: flex;
}
.cart-item-qty {
  display: flex;
}
.cart-item-remove-button {
  display: flex;
  margin-left: 10px;
}
