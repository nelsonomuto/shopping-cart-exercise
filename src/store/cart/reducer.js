import actions from "./actions"

const TAX_RATE = 0.1
const SHIPPING_COST_PER_ITEM = 5
const SHIPPING_COST_FLAT = 50
const SHIPPING_COST_RATE = 0.02

// Look up a product in the shopping cart by SKU
const findItem = (product, cartItems) =>
  cartItems.find(_item => _item.sku === product.sku)

const addProductToCart = (state, { product }) => {
  const foundItem = findItem(product, state.items)
  // if item exists, update the qty
  if (foundItem) {
    return incrementQuantity(state, { product })
  }
  // if new cart item, use qty 1
  const newItem = {
    ...product,
    qty: 1,
  }
  return {
    ...state,
    items: state.items.concat(newItem),
  }
}

const removeProductFromCart = (state, { product }) => {
  const foundItem = findItem(product, state.items)
  if (!foundItem) return state

  // item exists; remove it from the saved items
  const nextItems = state.items.filter(_item => _item !== foundItem)
  return {
    ...state,
    items: nextItems,
  }
}

const clearCart = state => {
  return {
    ...state,
    items: [],
  }
}

const updateCartItem = (state, foundItem, updates) => {
  const nextItem = {
    ...foundItem,
    ...updates,
  }
  const currentIndex = state.items.indexOf(foundItem)
  const nextItems = [...state.items]
  nextItems[currentIndex] = nextItem
  return {
    ...state,
    items: nextItems,
  }
}

const incrementQuantity = (state, { product }) => {
  const foundItem = findItem(product, state.items)
  if (!foundItem) return state
  return updateCartItem(state, foundItem, { qty: foundItem.qty + 1 })
}

const decrementQuantity = (state, { product }) => {
  const foundItem = findItem(product, state.items)
  if (!foundItem) return state
  const nextQty = foundItem.qty - 1
  if (nextQty <= 0) return removeProductFromCart(state, { product })
  return updateCartItem(state, foundItem, { qty: nextQty })
}

const toggleCartOpen = state => {
  return {
    ...state,
    isCartOpen: !state.isCartOpen,
  }
}

const reducer = (state, action) => {
  const _state = (() => {
    switch (action.type) {
      case actions.INCREMENT_QUANTITY:
        return incrementQuantity(state, action)
      case actions.DECREMENT_QUANTITY:
        return decrementQuantity(state, action)
      case actions.ADD_PRODUCT_TO_CART:
        return addProductToCart(state, action)
      case actions.REMOVE_PRODUCT_FROM_CART:
        return removeProductFromCart(state, action)
      case actions.CLEAR_CART:
        return clearCart(state)
      case actions.TOGGLE_CART_OPEN:
        return toggleCartOpen(state)
      default:
        return state
    }
  })()

  return {
    ..._state,
    summary: getCartSummary(_state.items),
  }
}

export const getCartSummary = cartItems => {
  const subtotal = cartItems.reduce(
    (acc, value) => acc + value.price * value.qty,
    0
  )
  const totalTax = subtotal * TAX_RATE
  const totalQuantity = cartItems.reduce((acc, value) => acc + value.qty, 0)
  const calculatedShipping =
    SHIPPING_COST_FLAT +
    subtotal * SHIPPING_COST_RATE -
    SHIPPING_COST_PER_ITEM * totalQuantity
  const shippingCost = calculatedShipping > 0 ? calculatedShipping : 0
  const totalCost = subtotal + totalTax + shippingCost

  return {
    subtotal,
    totalTax,
    shippingCost,
    totalCost,
  }
}

export default reducer
