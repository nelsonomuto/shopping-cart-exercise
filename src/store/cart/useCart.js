import createContextStore from "../context-reducer-store"
import reducer, { getCartSummary } from "./reducer"

const defaultState = {
  items: [],
  summary: getCartSummary([]),
}

const {
  store,
  StoreProvider: CartStoreProvider,
  useStore: useCart,
} = createContextStore("Cart", reducer, defaultState, {
  // only save items to localstorage, we dont need to save the whole cart state
  serializeStore: state => {
    return { items: state.items }
  },
  deserializeStore: state => {
    return {
      items: state.items,
      summary: getCartSummary(state.items),
    }
  },
})

export { store }
export { CartStoreProvider }
export { useCart }
