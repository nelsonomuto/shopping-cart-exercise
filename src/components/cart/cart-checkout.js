import React from "react"
import { dollars } from "../../services/formatString"
import { useCart } from "../../store/cart/useCart"
import { clearCart } from "../../store/cart/actions"
import ConfirmationButton from "../confirmation-button"
import CartList from "./cart-list"
import CartSummary from "./cart-summary"

const CartCheckout = () => {
  const [{ items, summary }, dispatch] = useCart()
  return (
    <>
      <CartList />
      <CartSummary />
      {items.length > 0 ? (
        <ConfirmationButton
          className="confirm-order-button"
          confirmText={`Confirm order for ${dollars(summary.totalCost)}`}
          cancelText="Cancel"
          onConfirm={() => {
            alert("Your order has been confirmed!")
            dispatch(clearCart())
          }}
        >
          Place Order
        </ConfirmationButton>
      ) : null}
    </>
  )
}

export default CartCheckout
