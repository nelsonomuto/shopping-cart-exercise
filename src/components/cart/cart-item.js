import React from "react"
import { dollars } from "../../services/formatString"
import {
  decrementQuantity,
  incrementQuantity,
  removeProductFromCart,
} from "../../store/cart/actions"
import { useCart } from "../../store/cart/useCart"

const CartItem = ({ item }) => {
  const [, dispatch] = useCart()
  const onRemove = () => dispatch(removeProductFromCart(item))
  const onIncrement = () => dispatch(incrementQuantity(item))
  const onDecrement = () => dispatch(decrementQuantity(item))

  return (
    <div className="cart-item">
      <span className="cart-item-title">{item.title}</span>
      <span className="cart-item-price">{dollars(item.price)}</span>
      <div className="cart-item-controls">
        <button className="cart-item-update-qty-button" onClick={onDecrement}>
          -
        </button>
        <span className="cart-item-qty">{item.qty}</span>
        <button className="cart-item-update-qty-button" onClick={onIncrement}>
          +
        </button>
        <button className="cart-item-remove-button" onClick={onRemove}>
          Remove Item
        </button>
      </div>
    </div>
  )
}

export default CartItem
