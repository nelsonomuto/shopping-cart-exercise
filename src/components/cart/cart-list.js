import { Link } from "gatsby"
import React from "react"
import { clearCart } from "../../store/cart/actions"
import { useCart } from "../../store/cart/useCart"
import CartItem from "./cart-item"

const CartList = () => {
  const [{ items }, dispatch] = useCart()

  if (items.length === 0) {
    return (
      <div className="cart-list cart-list-empty">
        <h3>
          Your cart is empty. <Link to="/">Shop now</Link>
        </h3>
      </div>
    )
  }

  return (
    <div className="cart-list">
      <ul>
        {items.map(item => (
          <li key={item.sku}>
            <CartItem item={item} />
          </li>
        ))}
      </ul>
      <button
        className="cart-list-clear-button"
        onClick={() => dispatch(clearCart())}
      >
        Clear Cart
      </button>
    </div>
  )
}

export default CartList
