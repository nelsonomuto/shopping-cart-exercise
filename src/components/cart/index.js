import { Link } from "gatsby"
import React, { useEffect, useRef } from "react"
import { dollars } from "../../services/formatString"
import { toggleCartOpen } from "../../store/cart/actions"
import { useCart } from "../../store/cart/useCart"
import CartList from "./cart-list"
import CartPortal from "./cart-portal"

const Cart = () => {
  const [{ isCartOpen, summary }, dispatch] = useCart()
  const containerRef = useRef(null)

  const toggleCart = () => dispatch(toggleCartOpen())

  const onClickOutside = e => {
    if (containerRef.current.contains(e.target)) {
      // inside
      return
    }
    // outside
    toggleCart()
  }

  useEffect(() => {
    if (isCartOpen) {
      document.addEventListener("mousedown", onClickOutside)
      return () => {
        document.removeEventListener("mousedown", onClickOutside)
      }
    } else {
      document.removeEventListener("mousedown", onClickOutside)
    }
  }, [isCartOpen])

  return (
    <CartPortal>
      <div
        onClick={onClickOutside}
        className={
          "cart-portal-container" +
          (isCartOpen ? " cart-animate-open" : " cart-animate-closed")
        }
      >
        <div className="cart-sidebar-container" ref={containerRef}>
          <button className="cart-portal-close-button" onClick={toggleCart}>
            Close
          </button>
          <div className="cart-sidebar-items-container">
            <h1>Cart</h1>
            <CartList onShop={toggleCart} />
          </div>
          <div className="cart-sidebar-summary-container">
            <div className="cart-subtotal">
              Subtotal: {dollars(summary.subtotal)}
            </div>
            <Link to="/checkout" className="cart-checkout-button">
              Checkout
            </Link>
          </div>
        </div>
      </div>
    </CartPortal>
  )
}

export default Cart
