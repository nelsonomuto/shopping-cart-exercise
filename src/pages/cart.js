import React from "react"
import "../assets/css/cart.css"
import CartList from "../components/cart/cart-list"
import CartSummary from "../components/cart/cart-summary"
import Layout from "../components/layout"
import SEO from "../components/seo"
import { CartStoreProvider } from "../store/cart/useCart"

const Cart = () => (
  <CartStoreProvider>
    <Layout>
      <SEO title="Cart" />
      <h1>Cart</h1>
      <CartList />
      <CartSummary />
    </Layout>
  </CartStoreProvider>
)

export default Cart
