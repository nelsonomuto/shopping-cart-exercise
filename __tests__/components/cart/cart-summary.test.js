import { shallow } from "enzyme"
import React from "react"
import CartSummary from "../../../src/components/cart/cart-summary"
import { getCartSummary } from "../../../src/store/cart/reducer"
import { useCart } from "../../../src/store/cart/useCart"

const mockItem = {
  sku: "fake-sku",
  title: "fake-title",
  price: 999.99,
  image: "fake-image-url",
  qty: 2,
}

// Mock useCart which is required in the component.
// Don't test that side effects from useCart actually happen,
// just test that dispatch is called where needed and that
// the original state gets rendered.

jest.mock("../../../src/store/cart/useCart.js", () => ({
  useCart: jest.fn(),
}))

describe("<CartSummary />", () => {
  it("matches snapshot", () => {
    useCart.mockImplementationOnce(() => [
      {
        items: [mockItem],
        summary: getCartSummary([mockItem]),
      },
    ])
    const component = shallow(<CartSummary />)
    expect(component).toMatchSnapshot()
  })

  it("is empty when items are empty", () => {
    useCart.mockImplementationOnce(() => [
      {
        items: [],
        summary: getCartSummary([]),
      },
    ])
    const component = shallow(<CartSummary />)
    expect(component).toEqual({})
  })

  it("shows free shipping when shipping cost is 0", () => {
    useCart.mockImplementationOnce(() => [
      {
        items: [mockItem],
        summary: {
          ...getCartSummary([mockItem]),
          shippingCost: 0,
        },
      },
    ])
    const component = shallow(<CartSummary />)
    expect(component).toMatchSnapshot()
  })
})
