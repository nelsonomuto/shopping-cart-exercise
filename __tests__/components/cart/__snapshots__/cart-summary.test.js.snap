// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartSummary /> matches snapshot 1`] = `
<div
  className="cart-summary"
>
  <div>
    Subtotal: 
    $ 1,999.98
  </div>
  <div>
    Tax: 
    $ 200.00
  </div>
  <div>
    Shipping:
     
    $ 80.00
  </div>
  <div>
    Total: 
    $ 2,279.98
  </div>
</div>
`;

exports[`<CartSummary /> shows free shipping when shipping cost is 0 1`] = `
<div
  className="cart-summary"
>
  <div>
    Subtotal: 
    $ 1,999.98
  </div>
  <div>
    Tax: 
    $ 200.00
  </div>
  <div>
    Shipping:
     
    FREE
  </div>
  <div>
    Total: 
    $ 2,279.98
  </div>
</div>
`;
